### PR TITLE
resource/aws_emrcontainers_virtual_cluster: new resource

### PR DESCRIPTION
**Pull request by shuheiktgw**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates https://github.com/hashicorp/terraform-provider-aws/issues/16717

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### CODE REVIEW

File: aws/resource_aws_emrcontainers_virtual_cluster.go [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+							ValidateFunc: validation.StringInSlice(emrcontainers.ContainerProviderType_Values(), false),
+						},
+					},
+				},
+			},
+			"created_at": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"id": {
```
Comment (DrFaust92) [inline aws/resource_aws_emrcontainers_virtual_cluster.go:84]: Not needed as TF always has a top level id attribute

File: aws/resource_aws_emrcontainers_virtual_cluster.go [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"name": {
```
Comment (DrFaust92) [inline aws/resource_aws_emrcontainers_virtual_cluster.go:88]: Lets add validation to `name` based on docs: https://docs.aws.amazon.com/emr-on-eks/latest/APIReference/API_CreateVirtualCluster.html#API_CreateVirtualCluster_RequestSyntax

Len is between 1 and 64 
and must conform to the following regex: `[\.\-_/#A-Za-z0-9]+`

File: aws/resource_aws_emrcontainers_virtual_cluster.go [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+		Name: aws.String(d.Get("name").(string)),
+	}
+
+	log.Printf("[INFO] Creating EMR containers virtual cluster: %s", input)
+	out, err := conn.CreateVirtualCluster(&input)
+	if err != nil {
+		return fmt.Errorf("error creating EMR containers virtual cluster: %w", err)
+	}
+
+	if _, err := waiter.VirtualClusterCreated(conn, aws.StringValue(out.Id)); err != nil {
```
Comment (DrFaust92) [inline aws/resource_aws_emrcontainers_virtual_cluster.go:115]: Missing SetId to the cluster id

**Review by DrFaust92 — CHANGES_REQUESTED** [review]

File: aws/resource_aws_emrcontainers_virtual_cluster.go (hunk 1 above)
Comment (shuheiktgw) [reply aws/resource_aws_emrcontainers_virtual_cluster.go:84]: Didn't know that! Thank you!

**Comment by shuheiktgw** [comment]
Hmm, while writing acc tests, I noticed those two steps are not possible through the current terraform-aws-provider:

- https://docs.aws.amazon.com/emr/latest/EMR-on-EKS-DevelopmentGuide/setting-up-cluster-access.html
- https://docs.aws.amazon.com/emr/latest/EMR-on-EKS-DevelopmentGuide/setting-up-enable-IAM.html

1. We cannot create Kubernetes resources from terraform-aws-provider
2. terraform-aws-provider still hasn't support EKS addons, which is necessary for IAM Roles for service accounts (https://github.com/hashicorp/terraform-provider-aws/issues/16542)

The second problem will be solved eventually but I'm not sure about how I can solve the first one... @DrFaust92 Do you have any ideas? 🙏

**Comment by DrFaust92** [comment]
Hey @shuheiktgw, do acc tests fail without those prerequisites? I don’t how they block creating the resource and using additional resources (and providers) to add access outside the scope of the tests

**Comment by shuheiktgw** [comment]
@DrFaust92 Yes without them, the acc test fails with the following error:

```
make testacc TESTARGS='-run=TestAccAwsEMRContainersVirtualCluster_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsEMRContainersVirtualCluster_basic -timeout 120m
=== RUN   TestAccAwsEMRContainersVirtualCluster_basic
=== PAUSE TestAccAwsEMRContainersVirtualCluster_basic
=== CONT  TestAccAwsEMRContainersVirtualCluster_basic
    resource_aws_emrcontainers_virtual_cluster_test.go:65: Step 1/2 error: Error running apply:
        Error: error creating EMR containers virtual cluster: ValidationException: Unauthorized to perform read namespace on default
```

I've looked into eksctl's PR to enable EMR containers, and they seem to create Kubernetes resources...

- https://github.com/weaveworks/eksctl/pull/2913

**Comment by shuheiktgw** [comment]
@bflad @ewbankkit Do you people know how I can write an acc test that requires provisioning Kubernetes resources?

**Comment by bflad** [comment]
@shuheiktgw if it does not work by just including `provider "kubernetes" {}` in the configuration, it may be necessary to also use [`helper/resource.TestCase` type `ExternalProviders` field](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource#TestCase), however we do not currently have any examples of multiple provider testing in this codebase anymore. Attempting to configure the Kubernetes provider using values from other resources may also require two `TestStep`.

The maintainers are very behind on the large backlog of existing pull requests, so unfortunately we cannot provide more guidance ourselves or guarantee when this will be reviewed.

**Comment by shuheiktgw** [comment]
Thank you @bflad! With that information, I think I can start writing the acc tests 👍

**Comment by anggao** [comment]
@shuheiktgw do you have any updates on this PR :) looking forward to use it

**Comment by ewbankkit** [comment]
Superseded by https://github.com/hashicorp/terraform-provider-aws/pull/20003.

**Pull request closed without merging**